core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -377,6 +377,19 @@ void CudaInternal::initialize(cudaStream_t stream, bool manage_stream) {
     Kokkos::abort("Calling Cuda::initialize after Cuda::finalize is illegal\n");
   was_initialized = true;
 
+  // Check that the device associated with the stream matches cuda_device
+  CUcontext context;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuStreamGetCtx(stream, &context)));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuCtxPushCurrent(context)));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuCtxGetDevice(&m_cudaDev)));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(m_cudaDev));
```

Comment:
Couldn't find a better way of doing this (i.e. not using driver interface) but we should be ok as long as Kokkos::initialize was called. Otherwise there could be issues with the context not being created yet.